File: kernel/src/main/kotlin/ninja/blacknet/transaction/RefundHTLC.kt
/*
 * Copyright (c) 2018-2024 Pavel Vasin
 *
 * Licensed under the Jelurida Public License version 1.1
 * for the Blacknet Public Blockchain Platform (the "License");
 * you may not use this file except in compliance with the License.
 * See the LICENSE.txt file at the top-level directory of this distribution.
 */

package ninja.blacknet.transaction

import kotlinx.serialization.Serializable
import ninja.blacknet.contract.HashTimeLockContractId
import ninja.blacknet.core.*
import ninja.blacknet.crypto.Hash

/**
 * 償還合約
 */
@Serializable
class RefundHTLC(
        val id: HashTimeLockContractId
) : TxData {
    override fun processCoinImpl(tx: Transaction, hash: Hash, dataIndex: Int, coinTx: CoinTx): Status {
        val htlc = coinTx.getHTLC(id)
        if (htlc == null) {
            return Invalid("HTLC not found")
        }
        if (tx.from != htlc.from) {
            return Invalid("Invalid sender")
        }
        if (!htlc.timeLock.verify(htlc.height, htlc.time, coinTx.height(), coinTx.blockTime())) {
            return Invalid("Invalid time lock")
        }

        val account = coinTx.getAccount(tx.from)!!
        account.debit(coinTx.height(), htlc.amount)
        coinTx.setAccount(tx.from, account)
        coinTx.removeHTLC(id)
        return Accepted
    }

    fun involves(ids: Set<HashTimeLockContractId>) = ids.contains(id)
}

File: kernel/src/main/rust/transaction/refundhtlc.rs
/*
 * Copyright (c) 2018-2025 Pavel Vasin
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

use crate::blake2b::Hash;
use crate::error::{Error, Result};
use crate::transaction::{CoinTx, HashTimeLockContractId, Transaction, TxData};
use alloc::borrow::ToOwned;
use serde::{Deserialize, Serialize};

#[derive(Deserialize, Serialize)]
pub struct RefundHTLC {
    id: HashTimeLockContractId,
}

impl TxData for RefundHTLC {
    fn process_impl(
        &self,
        tx: Transaction,
        _hash: Hash,
        _data_index: u32,
        coin_tx: impl CoinTx,
    ) -> Result<()> {
        let htlc = coin_tx.get_htlc(self.id)?;
        if tx.from() != htlc.from {
            return Err(Error::Invalid("Invalid sender".to_owned()));
        }
        htlc.time_lock.verify(
            htlc.height,
            htlc.time,
            coin_tx.height(),
            coin_tx.block_time(),
        )?;

        let mut account = coin_tx.get_account(tx.from())?;
        account.debit(coin_tx.height(), htlc.amount);
        coin_tx.set_account(tx.from(), account);
        coin_tx.remove_htlc(self.id);
        Ok(())
    }
}
